### DISCUSSION TITLE
Does QueryBuilder allow the input field to be set as an autocomplete select input field ?

### DISCUSSION THREAD

**Discussion by keerthankrishna** [react-querybuilder/react-querybuilder#430] [Q&A] [answered]
I am looking for a select input field which will fetch data from a JSON and will also be an autocomplete select text input - I'm not sure if this is possible looking at the documentation provided but wanted to reach out.

**Example:**
1. I have a _field_ called **instruments**
2. If I select that and if I select "**in**" under the "_Operator_" selection
3. Now, for the "_value_" field I am wanting it to be a text input but if i type "_pi_".. then it should show me all the input values that have a "_pi_" text.. like "_piano_". Similarly if i just type in "p" then i would need to get "_timpani_", "_piano_", "_trumpet_", "_xylophone_", "_saxophone_"

**Comment by jakeboone02** [comment, marked as answer]
There are a few autocomplete components out there that you can use with this library. I generally use [`react-select`](https://react-select.com/), either the standard `Select` component or the `Async` component if I need to load options asynchronously from an API.

Here is an example of a custom `ValueEditor` that uses `react-select` when the field is "autocomplete" and [falls back to the default `ValueEditor`](https://react-querybuilder.js.org/docs/tips/custom-with-fallback). The list fakes a delay for an API call with `setTimeout` and then loads the options that match what the user has typed.

https://codesandbox.io/s/broken-snow-b78i8f?file=/src/MyValueEditor.tsx

```tsx
import { ValueEditor, ValueEditorProps } from 'react-querybuilder';
import type { GroupBase, OptionsOrGroups } from 'react-select';
import AsyncSelect from 'react-select/async';

const options = [
  { value: 'v1', label: 'abc' },
  { value: 'v2', label: 'def' },
  { value: 'v3', label: 'ghi' },
  { value: 'v4', label: 'jkl' },
  { value: 'v5', label: 'mno' },
  { value: 'v6', label: 'pqr' },
  { value: 'v7', label: 'stu' },
  { value: 'v8', label: 'vwx' },
  { value: 'v9', label: 'yz' },
];

const loadDelay = 500;

const loadOptions = (
  iv: string
): Promise<OptionsOrGroups<any, GroupBase<any>>> =>
  new Promise(r =>
    setTimeout(() => r(options.filter(op => op.label.includes(iv))), loadDelay)
  );

export const MyValueEditor = (props: ValueEditorProps) => {
  if (props.field === 'autocomplete') {
    return (
      <AsyncSelect
        loadOptions={loadOptions}
        onChange={v => props.handleOnChange(v)}
        value={props.value}
        options={options}
      />
    );
  }

  return <ValueEditor {...props} />;
};
```

**Reply by keerthankrishna** [reply]
Awesome! Will this also work for the **fields** dropdown where you select _First Name_ ?

**Reply by jakeboone02** [reply]
Yeah, you can use `react-select` in a custom `fieldSelector` component as well. Same concept as the value editor but no need for the "fallback" since it only does the selection.

Actually you can make a generic, custom `ValueSelector` component and use it for _all_ the selectors if you want. `combinatorSelector`, `operatorSelector`, etc.

**Reply by keerthankrishna** [reply]
Thank you so much - I got it working with a custom `fieldSelector` component

**Reply by keerthankrishna** [reply]
Is there a right way to do it here ? I did it and I get this warning:
_A component is changing a controlled input to be uncontrolled. This is likely caused by the value changing from a defined to undefined, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component._

**Here's how my custom component looks like :**
```
const MyFieldSelectorProd = (prop: FIeldSelectorProps) => (
<Select id="FieldData" className="abcd" name="FieldData" options={fieldJSON} isClearable=(true) placeholder="Type to Search"></Select> );
```

**Here's how my QueryBuilder component looks like:**
```
<QueryBuilder
fields={fieldJSON}
setQuery={function}
query={query}
independentCombinators={true}
combinator={Combinators}
onQueryChange={q => function(q)}}
controlElements={{fieldSelector: My FieldSelectorProps}}
/>
```

**Reply by jakeboone02** [reply]
Can you reproduce the issue in CodeSandbox? You can use this as a template: https://codesandbox.io/s/github/react-querybuilder/react-querybuilder/tree/main/examples/basic-ts?file=/src/App.tsx

**Reply by keerthankrishna** [reply]
Thanks for the template @jakeboone02 ! I got it working in CodeSandbox (here's the link: https://codesandbox.io/s/zealous-glade-nly386?file=/src/App.tsx) I think that's the right way, yes ?

I will need to see what I messed up in my local code. I'm even having an issue with the field selection where it seems to get reset once I select something and then select the operator. I will dig deeper and confirm here what is wrong in my local code.

**Reply by jakeboone02** [reply]
Here's a fork of your sandbox that I think fixes all the issues.

https://codesandbox.io/s/magical-voice-1e6uv3?file=/src/App.tsx:1079-1384

I made four changes:

- Changed the type of `initialQuery` from `RuleGroupType` to `RuleGroupTypeIC` because you're using `independentCombinators`
- Added a `getOptionValue` prop to `Select`
- Added a `value` prop to `Select`
- Added an `onChange` prop to `Select`

**Reply by keerthankrishna** [reply]
Thank you so much @jakeboone02 ! I’ll take a look at this later tonight.. thanks a ton for taking the time for all this.

Does the QueryBuilder work the same way in a react solution ? I figured your template was typescript and my local code was a react setup with only JS files. I’ll setup another CodeSandBox for react and see if I have any issues there.

**Reply by jakeboone02** [reply]
Yeah, it works the same without TypeScript. I usually use TS in examples because it's more expressive, but there's an equivalent JavaScript template [here](https://github.com/react-querybuilder/react-querybuilder/tree/main/examples/basic).

Each subfolder in the [/examples](https://github.com/react-querybuilder/react-querybuilder/tree/main/examples) folder has a README file with links to CodeSandbox and StackBlitz that automatically load that folder's code.

**Reply by keerthankrishna** [reply]
Thank you so much @jakeboone02 ! Everything is working correctly now. 

I have a question regarding the below line of code which you added within the ` select ` in your CodeSandBox link (https://codesandbox.io/s/magical-voice-1e6uv3?file=/src/App.tsx:1168-1196). What does ` getOptionValue ` do and what happens when it is not put in there ?
` getOptionValue={f => f.name} `

I also have an issue when using this onChange function ` handleOnChange ` - I get the below error when the name becomes NULL (_this happens when I clear selection in the_ ` react-select ` _dropdown_). How do I handle it ? _The function seems to be an inbuilt QueryBuilder function that is being used ?_

![image](https://user-images.githubusercontent.com/24373382/209905841-154e6a2b-6c4e-459b-9c3e-55edb7601d19.png)

**Reply by jakeboone02** [reply]
`getOptionValue` is a `react-select` prop that resolves the `options` elements to a unique string (see docs [here](https://react-select.com/props), sorry it doesn't allow direct linking to the prop). If it's not there then `react-select` doesn't know to look at the `name` property of each field; it will try to look at `value` instead.

I updated the codesandbox to allow for clearing the field selector. All I did was add optional chaining to the `getOptionValue` and `onChange` handlers so they wouldn't try to read a property from `null`.

```
    getOptionValue={f => f?.name}
    //                    ^
    onChange={v => props.handleOnChange(v?.name)}
    //                                   ^
```

**Reply by keerthankrishna** [reply]
Thank you so much @jakeboone02 ! Everything is working correctly now.

**Reply by jakeboone02** [reply]
Minor update, January 2024: RQB version 7 (currently `v7.0.0-alpha.5`) will accept `value` as the identifier property for option lists. `name` will be optional, but both will be present in all option list props passed down to subcomponents.

This will eliminate the need, for example, to use the `getOptionValue` prop in `react-select` implementations since it uses `value` by default.
